#' @title Gene essentiality and differential essentiality scoring.
#'
#' @description Calls on all MoPAC modules in order to identify top essential and differentially-essential genes. If the data is in FASTQ format, please run the read.FASTQ function beforehand.
#'
#' @param id an optional identifier for the analysis. All future output will be stored in a directory with this name. No output will be produced if not supplied. Defaults to \code{NULL}.
#' @param counts either the path to the file containing the read counts or a data frame containing the read counts. It should include a column named "sgRNA". The read counts may or may not be already normalized.
#' @param library sgRNA library. It may be a data frame or a file path. It must contain a column called \code{"sgRNA"} and another called \code{"Gene"}. It may also contain a \code{"Category"} column. Only necessary if the table of counts does not include a column called "Gene".
#' @param annotation a data frame containing the sample annotation information. Should contain the following columns: "Sample", "Condition", "Replicate" and "Control".
#' @param condition1 an optional name of first condition to be compared. Defaults to the first condition available.
#' @param condition2 an optional name of second condition to be compared. Defaults to the second condition available.
#' @param pseudocount numeric. An optional pseudo-count for log2 transformation. If \code{NULL}, default optimization is performed over the values 0.01, 0.05, 0.1, 0.15, 0.2, 0.25 and 0.3. Defaults to 0.15.
#' @param empirical.weights logical. Should empirical weight values be used for sgRNA weighted averaging? If \code{FALSE}, sgRNA-rank-based optimization will be performed. Defaults to \code{FALSE}.
#' @param read.filtered logical. Should the filtered list of control genes generated by the 2-tail RRA algorithm be used to optimize the regression parameters? Requires an identifier to be set, otherwise use \code{filtered.essential} and \code{filtered.nonessential} instead. Defaults to \code{FALSE}.
#' @param read.unenriched logical. Should the filtered list of unenriched genes generated by the 2-tail RRA algorithm be used to perform the normalization? Requires an identifier to be set, otherwise please use \code{unenriched} instead. Defaults to \code{FALSE}.
#' @param report logical. Should visualization reports be generated? Requires a pandoc installation. Defaults to \code{FALSE}.
#'
#' @details
#' If \code{id} is not \code{NULL}, the following files are produced in the directory with name equal to the identifier:
#' \itemize{
#'  \item{\code{annotation.csv} and \code{annotation.xlsx}, two annotation files.}
#'  \item{\code{library.txt}, containing the sgRNAs mapped to the library file.}
#'  \item{\code{counts.txt}, containing the read counts per sgRNA for each sample.}
#'  \item{\code{sgRNA_log2.txt}, the log2-transformed read counts normalized by sequencing depth.}
#'  \item{\code{sgRNA_lfc_reps.txt} and \code{sgRNA_lfc.txt}, the log-fold-changes before and after averaging the replicates.}
#'  \item{\code{qc_variance.txt}, \code{qc_weights.txt}, \code{qc_scaling.txt} and \code{qc_shifting.txt}, the optimized preliminary regression parameters employing the unfiltered list of control genes whose size is equal to the most frequent gene size in the dataset.}
#'  \item{\code{qc.pdf}, a report for the visualization of the quality assessment.}
#'  \item{\code{RRA.txt} files, one for each condition.}
#'  \item{\code{essential.txt}, \code{nonessential.txt}, \code{nondifferential.txt}, \code{enriched.txt} and \code{unenriched.txt}, the filtered lists of essential, nonessential, nondifferential, enriched and unenriched genes.}
#'  \item{\code{condition1_condition2_normalized_reps.txt}, containing the gene-level normalized scores for the replicates of both conditions.}
#'  \item{\code{condition1_condition2_differential.txt}, the relative difference statistic for the pair of conditions selected.}
#' }
#'
#' @return A data frame containing the MoPAC differential gene essentiality scores of the two selected conditions.
#'
#' @author Oscar D Villarreal, \email{oscardvillarreal@gmail.com}
#' @keywords mopac pipeline
#'
#' @references Li, W. et al. MAGeCK enables robust identification of essential genes from genome-scale CRISPR/Cas9 knockout screens. Genome Biol. 15, 554 (2014).
#'
#' @examples
#'
#' ## 1. The read counts must include a column named "sgRNA".
#' ## Either the path to the file containing the read counts or a data frame may be used:
#' counts <- MoPAC::dang_cck81
#' library <- MoPAC::sgRNA_library
#'
#' ## 2. Prepare the sample annotation data frame:
#' annotation <- data.frame(Sample=names(counts)[sapply(counts,is.numeric)])
#' annotation$Condition <- c("Plasmid0","Plasmid1",rep("DANG",4),rep("CCK81",3))
#' annotation$Replicate <- c("A","A","A","B","C","D","A","B","C")
#' annotation$Control <- c("","",rep("Plasmid0",4),rep("Plasmid1",3))
#'
#' ## 3. Compute essentiality and differential essentiality scores:
#'
#' ## a) If no output files are desired but rather just a data frame of the gene scores:
#' result <- mopac(counts=counts, library=library, annotation=annotation, pseudocount=0.15,
#'                 empirical.weights=TRUE, read.unenriched=TRUE, report=FALSE)
#'
#' ## b) Output files and pdf/html reports can be produced by simply adding an identifier:
#' result <- mopac(id="DANG_CCK81", counts=counts, library=library, annotation=annotation,
#'                 pseudocount=0.15, empirical.weights=TRUE, read.unenriched=TRUE,
#'                 report=FALSE)
#'
#' ## c) If there are more than 2 conditions, 2 of them must be specified for comparison.
#' result <- mopac(id="DANG_CCK81", counts=counts, library=library, annotation=annotation,
#'                 pseudocount=0.15, empirical.weights=TRUE, read.unenriched=TRUE,
#'                 condition1="DANG", condition2="CCK81", report=FALSE)
#'
#' ##--------------------------------------------------------------------------------------
#'
#' @export
#'

mopac <- function(id=NULL,counts=NULL,library=NULL,annotation=NULL,condition1=NULL,condition2=NULL,pseudocount=NULL,empirical.weights=FALSE,read.filtered=FALSE,read.unenriched=FALSE,report=FALSE) {
  set.seed(123)

  # Read input:
  print("Reading input...")
  counts <- read.counts(id=id,library=library,counts=counts,annotation=annotation)

  # Pre-processing and quality control:
  print("Pre-processing...")
  qc <- quality.control(id=id,counts=counts$Counts,annotation=counts$Annotation,pseudocount=pseudocount,report=report)

  # 2-tail RRA essentiality analysis, with or without control gene filtering and nondifferential gene extraction:
  if(read.filtered==TRUE | read.unenriched==TRUE) {
    print("Starting two-tail RRA analysis...")
    rra <- RRA.2tail(id=id,sgRNA=qc$sgRNA_lfc,conditions=c(condition1,condition2),pvalue=0.05,fraction=0.8)
  } else rra <- NULL
  if(read.filtered==TRUE) {
    filtered.essential <- rra$Essential
    filtered.nonessential <- rra$Nonessential
  } else {
    filtered.essential <- NULL
    filtered.nonessential <- NULL
  }
  if(read.unenriched==TRUE) {
    unenriched <- rra$Unenriched
  } else unenriched <- NULL

  # MoPAC gene essentiality analysis:
  print("Calculating MoPAC essentiality scores...")
  essentiality <- analyze.essentiality(id=id,sgRNA_reps=qc$sgRNA_lfc_reps,annotation=counts$Annotation,empirical.weights=empirical.weights,read.filtered=read.filtered,filtered.essential=filtered.essential,filtered.nonessential=filtered.nonessential)

  # MoPAC differential gene essentiality analysis:
  conditions <- unique(counts$Annotation$Condition[!counts$Annotation$Condition%in%counts$Annotation$Control])
  if(length(conditions)>1) {
    print("Calculating MoPAC differential essentiality scores...")
    differential <- analyze.differential(id=id,sgRNA=qc$sgRNA_lfc,scored_reps=essentiality$scored_reps,annotation=counts$Annotation,condition1=condition1,condition2=condition2,read.unenriched=read.unenriched,unenriched=unenriched,report=report)
    return(list(Counts=counts,QC=qc,RRA=rra,Essentiality=essentiality,Differential=differential))
  } else {
    print("The data contains a single condition. Returning essentiality scores...")
    return(list(Counts=counts,QC=qc,RRA=rra,Essentiality=essentiality))
  }
}


